src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs

```diff
@@ -122,11 +117,115 @@ protected virtual LabelSymbol GetDagNodeLabel(BoundDecisionDagNode dag)
+                }
+
+                public bool MightAssignSomething(BoundExpression expr)
+                {
+                    if (expr == null || expr.ConstantValue != null)
+                    {
+                        return false;
+                    }
+
+                    this._mightAssignSomething = false;
+                    this.Visit(expr);
+                    return this._mightAssignSomething;
+                }
+
+                public override BoundNode VisitCall(BoundCall node)
+                {
+                    _mightAssignSomething =
+                        // might be a call to a local function that assigns a pattern temp
+                        _isSwitchStatement && node.Method.MethodKind == MethodKind.LocalFunction ||
+                        // or perhaps we are passing a variable by ref and mutating it that way
+                        !node.ArgumentRefKindsOpt.IsDefault;
+                    return base.VisitCall(node);
+                }
+
+                public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
+                {
+                    _mightAssignSomething = true;
+                    return null;
+                }
+
+                public override BoundNode VisitCompoundAssignmentOperator(BoundCompoundAssignmentOperator node)
+                {
+                    _mightAssignSomething = true;
+                    return null;
+                }
+
+                public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstructionAssignmentOperator node)
+                {
+                    _mightAssignSomething = true;
+                    return null;
+                }
+
+                public override BoundNode VisitObjectCreationExpression(BoundObjectCreationExpression node)
+                {
+                    // perhaps we are passing a variable by ref and mutating it that way
+                    _mightAssignSomething = !node.ArgumentRefKindsOpt.IsDefault;
+                    return base.VisitObjectCreationExpression(node);
+                }
+            }
+            protected BoundDecisionDag ShareTempsIfPossibleAndEvaluateInput(
+                BoundDecisionDag decisionDag,
+                BoundExpression loweredSwitchGoverningExpression,
+                ArrayBuilder<BoundStatement> result)
+            {
+                // Note that a when-clause can contain an assignment to a
+                // pattern variable declared in a different when-clause (e.g. in the same section, or
+                // in a different section via the use of a local function), so we need to analyze all
+                // of the when clauses to see if they are all simple enough to conclude that they do
+                // not mutate pattern variables.
+                var mightAssignWalker = new WhenClauseMightAssignWalker(isSwitchStatement: this._isSwitchStatement);
+                bool canShareTemps =
```

Comment:
>bool canShareTemps [](start = 16, length = 18)

Consider merging `OfType` and `Any` to avoid `IEnumerable` allocations:
```
!decisionDag.TopologicallySortedNodes
.Any(n => n as BoundWhenDecisionDagNode w &&
    mightAssignWalker.MightAssignSomething(w.WhenExpression));
```